src/main.rs

```diff
@@ -414,31 +403,9 @@ fn main() {
     }
 
     // Some flags can be run without a command.
-    if !(args.flag_update || args.flag_clear_cache || args.flag_config_path) {
-        eprintln!("{}", USAGE);
+    if !(args.update || args.clear_cache || args.config_path) {
```

Comment:
@pksunkara how would I implement this logic in `clap_derive`?

In structopt, it seems that for custom help printing I could simply use [`args.clap().print_help()`](https://docs.rs/structopt/0.3.11/structopt/trait.StructOpt.html?search=#tymethod.clap) but that doesn't seem to be available in clap itself.

Reply:
I don't get this part. Are you trying to print help if no other arguments are given?

Reply:
Some flags (e.g. `--update`, `--clear-cache` or `--config-path`) can be used without a command. Other flags (e.g. `--pager` or `--os`) can only be used if a command is provided.

The command is a TLDR page in this case, so we can't create a finite list of clap subcommands.

Reply:
```
#[derive(Clap, Debug)]
#[clap(about = "A fast TLDR client", author, version)]
struct Args {
    /// The command to show (e.g. `tar` or `git log`)
    command: Vec<String>,

    /// Use a pager to page output
    #[clap(short = "p", long = "pager", requires = "command")]
    pager: bool,

    /// Show config file path
    #[clap(long = "config-path")]
    config_path: bool,
}
```

Clap v2 docs should mostly still be correct regarding function names. https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.requires

Also you don't need the verison option.

Reply:
> Clap v2 docs should mostly still be correct regarding function names.

Ah, I see! Will try to implement those dependencies.

> Also you don't need the verison option.

I need it if I want the short argument to be `-v` instead of `-V`, right?

Reply:
True yeah. I didn't realize you had it short.

Reply:
@pksunkara: I tried this:

```rust
#[derive(Clap, Debug)]
#[clap(about = "A fast TLDR client", author, version)]
struct Args {
    /// The command to show (e.g. `tar` or `git log`)
    command: Vec<String>,
...
    /// Display the raw markdown instead of rendering it
    #[clap(short = "m", long = "markdown", requires = "command")]
    markdown: bool,
...
```

However, that `requires` doesn't seem to work as intended because `command` is a Vec, so an empty command list is still an empty command list. I'd need something like "requires that `command` is not empty".

Reply:
`Option<Vec<String>>` should work IIRC.

Reply:
If not, try `Vec<string>` with `min_values = 1`

Docs regarding those types, https://docs.rs/structopt/0.3.11/structopt/#type-magic. Mostly still be same to the clap_derive.

Reply:
Clap doesn't like `Option<Vec<_>>`, since it doesn't know whether a missing argument should be `None` or `Some(vec![])`. However, `min_values = 1` seems to work beautifully :slightly_smiling_face:

Reply:
Please do keep me up to date on the perf/size/valgrind improvements you notice.

Reply:
Ok, I pushed an update. Two issues I noticed:

- Even though I set `min_values = 1` on `command`, when I run the binary without any arguments, it still runs but doesn't print any output. In that case, args are: `Args { command: [], list: false, render: None, os: None, update: false, clear_cache: false, pager: false, quiet: false, markdown: false, config_path: false, seed_config: false, version: false }` Can I make the command required somehow unless a flag that doesn't require it is used?

- I added a conflict between `command` and `--render`:

    ```diff
         /// Render a specific markdown file
    -    #[clap(short = "f", long = "render")]
    +    #[clap(short = "f", long = "render", conflicts_with = "command")]
         render: Option<String>,
    ```

  However, when padding in both a command and the `--render` flag, I get this error message:

   ```
     [clap]danilo@x2000:tealdeer$ cargo run -- --render tar.md gz
       Finished dev [unoptimized + debuginfo] target(s) in 0.05s
          Running `target/debug/tldr --render tar.md gz`
    error: The argument '--render <render>' cannot be used with '--render <render>'
    
    USAGE:
        tldr [FLAGS] [OPTIONS] [command]...
    
    For more information try --help
    ```

    Bug?

Reply:
https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.required

Can you move `command` to the end of the struct? It might fix the bug. `command` here is a positional argument here which means most people generally put it at the end. Either way, please create an issue on clap repo.

Reply:
My bad, wrong link.  Check https://docs.rs/clap/2.33.0/clap/struct.Arg.html#method.required_unless and other functions.

Reply:
Ah, the bug is https://github.com/clap-rs/clap/issues/1622! Reordering the struct does not help.

Reply:
So, that bug is being fixed today.

Reply:
Just saw it, awesome :slightly_smiling_face: